File: mobile-engage/src/main/java/com/emarsys/mobileengage/request/CoreCompletionHandlerRefreshTokenProxy.java
package com.emarsys.mobileengage.request;

import androidx.annotation.Nullable;

import com.emarsys.core.CoreCompletionHandler;
import com.emarsys.core.api.result.CompletionListener;
import com.emarsys.core.endpoint.ServiceEndpointProvider;
import com.emarsys.core.request.RestClient;
import com.emarsys.core.response.ResponseModel;
import com.emarsys.core.storage.Storage;
import com.emarsys.core.util.Assert;
import com.emarsys.mobileengage.RefreshTokenInternal;
import com.emarsys.mobileengage.util.RequestModelUtils;

import java.util.Objects;


public class CoreCompletionHandlerRefreshTokenProxy implements CoreCompletionHandler {
    private final CoreCompletionHandler coreCompletionHandler;
    private final RefreshTokenInternal refreshTokenInternal;
    private final RestClient restClient;
    private final Storage<String> contactTokenStorage;
    private final ServiceEndpointProvider clientServiceProvider;
    private final ServiceEndpointProvider eventServiceProvider;
    private final ServiceEndpointProvider messageInboxServiceProvider;

    public CoreCompletionHandlerRefreshTokenProxy(CoreCompletionHandler coreCompletionHandler, RefreshTokenInternal refreshTokenInternal, RestClient restClient, Storage<String> contactTokenStorage, ServiceEndpointProvider clientServiceProvider, ServiceEndpointProvider eventServiceProvider, ServiceEndpointProvider messageInboxServiceProvider) {
        Assert.notNull(coreCompletionHandler, "CoreCompletionHandler must not be null!");
        Assert.notNull(refreshTokenInternal, "RefreshTokenInternal must not be null!");
        Assert.notNull(restClient, "RestClient must not be null!");
        Assert.notNull(contactTokenStorage, "ContactTokenStorage must not be null!");
        Assert.notNull(clientServiceProvider, "ClientServiceProvider must not be null!");
        Assert.notNull(eventServiceProvider, "EventServiceProvider must not be null!");
        Assert.notNull(messageInboxServiceProvider, "MessageInboxServiceProvider must not be null!");

        this.coreCompletionHandler = coreCompletionHandler;
        this.refreshTokenInternal = refreshTokenInternal;
        this.restClient = restClient;
        this.contactTokenStorage = contactTokenStorage;
        this.clientServiceProvider = clientServiceProvider;
        this.eventServiceProvider = eventServiceProvider;
        this.messageInboxServiceProvider = messageInboxServiceProvider;
    }

    @Override
    public void onSuccess(String id, ResponseModel responseModel) {
        coreCompletionHandler.onSuccess(id, responseModel);
    }

    @Override
    public void onError(final String originalId, final ResponseModel originalResponseModel) {
        if (originalResponseModel.getStatusCode() == 401 && RequestModelUtils.isMobileEngageV3Request(originalResponseModel.getRequestModel(), eventServiceProvider, clientServiceProvider, messageInboxServiceProvider)) {
            refreshTokenInternal.refreshContactToken(new CompletionListener() {
                @Override
                public void onCompleted(@Nullable Throwable errorCause) {
                    if (errorCause == null) {
                        restClient.execute(originalResponseModel.getRequestModel(), CoreCompletionHandlerRefreshTokenProxy.this);
                    } else {
                        for (String id : com.emarsys.core.util.RequestModelUtils.extractIdsFromCompositeRequestModel(originalResponseModel.getRequestModel())) {
                            coreCompletionHandler.onError(id, new Exception(errorCause));
                        }
                    }
                }
            });
        } else {
            coreCompletionHandler.onError(originalId, originalResponseModel);
        }
    }

    @Override
    public void onError(String id, Exception cause) {
        coreCompletionHandler.onError(id, cause);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        CoreCompletionHandlerRefreshTokenProxy that = (CoreCompletionHandlerRefreshTokenProxy) o;

        if (!Objects.equals(coreCompletionHandler, that.coreCompletionHandler))
            return false;
        if (!Objects.equals(refreshTokenInternal, that.refreshTokenInternal))
            return false;
        if (!Objects.equals(restClient, that.restClient))
            return false;
        return Objects.equals(contactTokenStorage, that.contactTokenStorage);
    }

    @Override
    public int hashCode() {
        int result = coreCompletionHandler != null ? coreCompletionHandler.hashCode() : 0;
        result = 31 * result + (refreshTokenInternal != null ? refreshTokenInternal.hashCode() : 0);
        result = 31 * result + (restClient != null ? restClient.hashCode() : 0);
        result = 31 * result + (contactTokenStorage != null ? contactTokenStorage.hashCode() : 0);
        return result;
    }
}

File: mobile-engage/src/main/java/com/emarsys/mobileengage/request/CoreCompletionHandlerRefreshTokenProxy.kt
package com.emarsys.mobileengage.request

import com.emarsys.core.CoreCompletionHandler
import com.emarsys.core.endpoint.ServiceEndpointProvider
import com.emarsys.core.request.RestClient
import com.emarsys.core.response.ResponseModel
import com.emarsys.core.storage.Storage
import com.emarsys.core.util.RequestModelUtils
import com.emarsys.mobileengage.RefreshTokenInternal
import com.emarsys.mobileengage.util.RequestModelUtils.isMobileEngageRequest

class CoreCompletionHandlerRefreshTokenProxy(private val coreCompletionHandler: CoreCompletionHandler,
                                             private val refreshTokenInternal: RefreshTokenInternal,
                                             private val restClient: RestClient,
                                             private val contactTokenStorage: Storage<String>,
                                             private val pushTokenStorage: Storage<String>) : CoreCompletionHandler {

    override fun onSuccess(id: String, responseModel: ResponseModel) {
        coreCompletionHandler.onSuccess(id, responseModel)
    }

    override fun onError(originalId: String, originalResponseModel: ResponseModel) {
        if (originalResponseModel.statusCode == 401
                && originalResponseModel.isMobileEngageRequest()) {
            pushTokenStorage.remove()
            refreshTokenInternal.refreshContactToken { errorCause ->
                if (errorCause == null) {
                    restClient.execute(originalResponseModel.requestModel, this)
                } else {
                    for (id in RequestModelUtils.extractIdsFromCompositeRequestModel(originalResponseModel.requestModel)) {
                        coreCompletionHandler.onError(id, Exception(errorCause))
                    }
                }
            }
        } else {
            coreCompletionHandler.onError(originalId, originalResponseModel)
        }
    }

    override fun onError(id: String, cause: Exception) {
        coreCompletionHandler.onError(id, cause)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as CoreCompletionHandlerRefreshTokenProxy

        if (coreCompletionHandler != other.coreCompletionHandler) return false
        if (refreshTokenInternal != other.refreshTokenInternal) return false
        if (restClient != other.restClient) return false
        if (contactTokenStorage != other.contactTokenStorage) return false
        if (pushTokenStorage != other.pushTokenStorage) return false

        return true
    }

    override fun hashCode(): Int {
        var result = coreCompletionHandler.hashCode()
        result = 31 * result + refreshTokenInternal.hashCode()
        result = 31 * result + restClient.hashCode()
        result = 31 * result + contactTokenStorage.hashCode()
        result = 31 * result + pushTokenStorage.hashCode()
        return result
    }

}